lib/DxilValidation/DxilContainerValidation.cpp

```diff
@@ -73,6 +78,522 @@ struct DiagRestore {
+      PCInputToOutputTable = std::vector<uint32_t>(AlignedTabSize, 0);
+      memcpy(PCInputToOutputTable.data(), &(Data[Offset]), TabSize * 4);
+      Offset += TabSize;
+    }
+  }
+  IsValid = Offset == Data.size();
+}
+
+class PSVContentVerifier {
+  DxilModule &DM;
+  DxilPipelineStateValidation &PSV;
+  ValidationContext &ValCtx;
+  bool PSVContentValid = true;
+
+public:
+  PSVContentVerifier(DxilPipelineStateValidation &PSV, DxilModule &DM,
+                     ValidationContext &ValCtx)
+      : DM(DM), PSV(PSV), ValCtx(ValCtx) {}
+  void Verify();
+
+private:
+  void VerifySignatures(unsigned ValMajor, unsigned ValMinor);
+  void VerifySignature(const DxilSignature &, PSVSignatureElement0 *Base,
+                       unsigned Count, std::string Name,
+                       bool i1ToUnknownCompat);
+  void VerifySignatureElement(const DxilSignatureElement &,
+                              PSVSignatureElement0 *, const PSVStringTable &,
+                              const PSVSemanticIndexTable &, std::string, bool);
+  void VerifyResources(unsigned PSVVersion);
+  template <typename T>
+  void VerifyResourceTable(T &ResTab, unsigned &ResourceIndex,
+                           unsigned PSVVersion);
+  void VerifyViewIDDependence(PSVRuntimeInfo1 *PSV1);
+  void VerifyEntryProperties(const ShaderModel *SM, PSVRuntimeInfo0 *PSV0,
+                             PSVRuntimeInfo1 *PSV1, PSVRuntimeInfo2 *PSV2);
+  void EmitMismatchError(StringRef Name, StringRef PartContent,
+                         StringRef ModuleContent) {
+    ValCtx.EmitFormatError(ValidationRule::ContainerContentMatches,
+                           {Name, "PSV0", PartContent, ModuleContent});
+    PSVContentValid = false;
+  }
+  void EmitInvalidError(StringRef Name) {
+    ValCtx.EmitFormatError(ValidationRule::ContainerContentInvalid, {Name});
+    PSVContentValid = false;
+  }
+  template <typename Ty> static std::string GetDump(const Ty &T) {
+    std::string Str;
+    raw_string_ostream OS(Str);
+    T.Print(OS);
+    OS.flush();
+    return Str;
+  }
+};
+
+bool ViewIDTableAndMaskMismatched(const PSVDependencyTable &PSVTable,
+                                  std::vector<uint32_t> &VecPSVTable,
+                                  const PSVComponentMask &&PSVMask,
+                                  MutableArrayRef<uint32_t> ArrayMask,
+                                  uint32_t NumInputVectors,
+                                  uint32_t NumOutputVectors, bool UsesViewID) {
+  if (NumInputVectors > 0 && NumOutputVectors > 0) {
+    const PSVDependencyTable DxilTable(VecPSVTable.data(), NumInputVectors,
+                                       NumOutputVectors);
+    if (PSVTable != DxilTable)
+      return true;
+  }
+
+  if (UsesViewID && NumOutputVectors > 0) {
+    PSVComponentMask DxilMask(ArrayMask.data(), NumOutputVectors);
+    if (PSVMask != DxilMask)
+      return true;
+  }
+  return false;
```

Comment:
Can you explain this a bit? IN cases where there are only input vectors, or only output vectors and ViewID isn't used, there isn't any check for matching at all. Is that okay? If these situations are not expected to happen, I think this should still return that they mismatched and/or assert.

Reply:
These situations are not expected to happen.
If the situation happened, the loader of PSV or builder of ViewIDState should catch it. And we'll go the invalid path.
I'll add assert.